http/src/v0/pin.rs

```diff
@@ -116,26 +107,224 @@ impl<'a> TryFrom<&'a str> for AddRequest {
+    Recursive,
+    All,
+}
+
+impl Default for PinFilter {
+    fn default() -> Self {
+        PinFilter::All
+    }
+}
+
+impl PinFilter {
+    fn to_mode(&self) -> Option<PinMode> {
+        use PinFilter::*;
+        match self {
+            Direct => Some(PinMode::Direct),
+            Indirect => Some(PinMode::Indirect),
+            Recursive => Some(PinMode::Recursive),
+            All => None,
+        }
+    }
+}
+
+/// `pin/ls` per https://docs.ipfs.io/reference/http/api/#api-v0-pin-ls
+pub fn list<T: IpfsTypes>(
+    ipfs: &Ipfs<T>,
+) -> impl Filter<Extract = (impl Reply,), Error = Rejection> + Clone {
+    with_ipfs(ipfs).and(list_options()).and_then(list_inner)
+}
+
+fn list_options() -> impl Filter<Extract = (ListRequest,), Error = Rejection> + Clone {
     warp::filters::query::raw().and_then(|q: String| {
-        let res = AddRequest::try_from(q.as_str())
+        let res = ListRequest::try_from(q.as_str())
             .map_err(StringError::from)
             .map_err(warp::reject::custom);
 
         futures::future::ready(res)
     })
 }
+
+async fn list_inner<T: IpfsTypes>(
+    ipfs: Ipfs<T>,
+    req: ListRequest,
+) -> Result<impl Reply, Rejection> {
+    use futures::stream::{StreamExt, TryStreamExt};
+
+    #[derive(serde::Serialize)]
+    struct Good {
+        #[serde(rename = "Cid")]
+        cid: StringSerialized<Cid>,
+        #[serde(rename = "Type")]
+        mode: Cow<'static, str>,
+    }
+
+    impl From<(Cid, Cow<'static, str>)> for Good {
+        fn from((cid, mode): (Cid, Cow<'static, str>)) -> Good {
+            Good {
+                cid: StringSerialized(cid),
+                mode,
+            }
+        }
+    }
+
+    if req.arg.is_empty() {
+        let st = ipfs.list_pins(req.filter.to_mode()).await;
+
+        if req.stream {
+            let st = st.map_ok(|(cid, mode)| {
+                Good::from((
+                    cid,
+                    match mode {
+                        PinMode::Direct => Cow::Borrowed("direct"),
+                        PinMode::Indirect => Cow::Borrowed("indirect"),
+                        PinMode::Recursive => Cow::Borrowed("recursive"),
+                    },
+                ))
+            });
+
+            Ok(format_json_newline(st))
+        } else {
+            // TODO: the non stream variant looks like the http docs one:
+            // { "Keys": { "cid": { "Type": "indirect" } } }
+            Err(crate::v0::NotImplemented.into())
+        }
+    } else {
+        // the variant where we are not actually listing anything but more about finding the
+        // specific cids
+
+        let requirement = req.filter.to_mode();
+
+        let details = ipfs
+            .query_pins(req.arg, requirement)
+            .await
+            .map_err(StringError::from)?;
+
+        if req.stream {
+            let st = futures::stream::iter(details)
+                .map(Ok::<_, std::convert::Infallible>) // only done trying to match the types
```

Comment:
sneaky, but I'm ok with this :ok_hand: :smiley:

Reply:
yeah I don't think there is any other way. Luckily Infallible is a good error.